### ISSUE TITLE
Duplicate keys are not handled properly

### ISSUE THREAD

**Issue by horenmar** [yaml/pyyaml#165]
[YAML spec in version 1.2](http://yaml.org/spec/1.2/spec.html) says that in a valid YAML file, mapping keys are unique.

This is not the case in pyyaml, as can be seen by loading this sample file:
```yaml
a:
  - b
  - c

a:
  q: a
  q: b
```

The correct result from loading this file is an error. pyyaml instead naively overwrites results with the last key, resulting in this dict: `{'a': {'q': 'b'}}`.

**Comment by anz-rfc** [comment]
some discussion of workarounds here: https://gist.github.com/pypt/94d747fe5180851196eb

**Comment by mady143** [comment]
I am also getting same issue but its slight different  click here: https://github.com/gunthercox/ChatterBot/issues/1640

**Comment by parrenin** [comment]
I just wanted to comment that in my case, the overwrite is a feature and not a bug.
In my model, I have several "sites", each of them with a yaml parameter file.
And I have also a parameter file which apply to all sites.
So I concatenate the general parameter file with each individual parameter file which thus overwrite the general parameter file when there is a duplicate entry.
So it would be helpful to at least keep an option for the overwrite feature.

**Comment by perlpunk** [comment]
One problem is that this logic would break the functionality of merge keys:
https://yaml.org/type/merge.html

To fix this, we would need a seperate handling of the merge keys and the actual keys in the mapping.

I don't know which tests were failing (haven't tested your patch yet), but I guess the ones with merge keys were among them.

**Comment by perlpunk** [comment]
@marvingreenberg yeah, I think forbidding duplicate keys should be the default.
Adding an option for disabling it might be nice.

But merge key behaviour should stay the same.

@ingydotnet is the one to decide about this.

**Comment by perlpunk** [comment]
@parrenin 
> So I concatenate the general parameter file with each individual parameter file which thus overwrite the general parameter file when there is a duplicate entry

I think this sounds like a use case for merge keys, actually.

**Comment by wimglenn** [comment]
I suppose PyYAML might be using the 1.1 spec where the mapping keys aren't required to be unique?  https://yaml.org/spec/1.1/current.pdf

**Comment by perlpunk** [comment]
The YAML 1.1 spec also says that keys should be unique: https://yaml.org/spec/1.1/#id861060

> The content of a mapping node is an unordered set of key: value node pairs, with the restriction that each of the keys is unique

**Comment by wimglenn** [comment]
@perlpunk  Does that include the merge keys themselves, or only the keys after a resolved merge?

e.g.

```
- k0: v0
  vars: &splat0
    VAR0: potato  # vars from blob0

- k1: v1
  vars: &splat1
    VAR1: spud    # vars from blob1

- k: v
  extra:          # merged vars from blob0 + blob1
    <<: *splat0
    <<: *splat1
```

**Comment by perlpunk** [comment]
@wimglenn the correct way to use two merge keys is actually:
```
- k: v
  extra:          # merged vars from blob0 + blob1
    <<: [ *splat0, *splat1 ]
```
That way duplicate merge keys can be avoided.

Other than that, the spec is actually not very specific about equality of nodes.
For example, `0xa` and `10` are different when parsed, but equal when resolved.

**Comment by wimglenn** [comment]
Then what about

```
- k: v
  extra:
    VAR0: before
    <<: [ *splat0, *splat1 ]
    VAR1: after
```

I know merge spec says that splat0 wins in case of duplicate keys in splat0 & splat1, but what does it say about dupes outside of the merge caused by resolving the merge?

**Comment by perlpunk** [comment]
@wimglenn Duplicates should only be considered in the actual mapping definition itself. Merge keys are explicitly there for merging, so they should *override* keys.

By the way, the spec https://yaml.org/type/merge.html is not very detailed about this, but from how I read the spec and the example, it does not matter where the merge key appers - as the first key, in the middle or at the end. The order is always the same:
```
---
map:
  <<: [ *alias2, *alias3, *alias4 ]
  key: value1
  # key: something # would be an error
```
`value1` overrides `*alias2`, `*alias2` overrides `*alias3`, etc.

**Comment by alexchandel** [comment]
PyYAML should reject duplicate keys per the YAML spec.  These are silent errors in users' code.  Do what `ruamel.yaml` does, and add a `yaml.allow_duplicate_keys = True` property for those who want to opt-in.

**Comment by perlpunk** [comment]
@alexchandel yes, but it's not that trivial to implement, if we want to keep merge key functionality as well. Like I said earlier https://github.com/yaml/pyyaml/issues/165#issuecomment-525047100

**Comment by Fleker** [comment]
The workaround that [anz-rfc](https://github.com/yaml/pyyaml/issues/165#issuecomment-430074049) posted works well enough for me, though a more canonical solution would be preferable.

**Comment by aclum** [comment]
bump

**Comment by lishaogang** [comment]
this code can really handle the duplicate key problem.
```
import yaml
import yaml.constructor
import yaml.resolver

def check_duplicate_keys(loader: yaml.SafeLoader, node):    
    ret =  loader.construct_mapping(node)
    keys = []
    for key_node, value_node in node.value:
        if key_node.value in keys:
            raise yaml.constructor.ConstructorError(
                f"Duplicate key '{key_node.value}' found",
                key_node.start_mark,
                "Each key must be unique in YAML mapping",
                key_node.start_mark
            )
        keys.append(key_node.value)
    
    return ret


yaml.SafeLoader.add_constructor(
    yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG,
    check_duplicate_keys
)


src_path = r'demo.yaml'
try:
    with open(src_path, 'r') as f:
        data = yaml.safe_load(f)
except yaml.constructor.ConstructorError as e:
    print(f"Duplicate key detected: {e}")
else:
    print(data)
```

valid input:
```
- k0: v0
  vars: &splat0
    VAR0: potato  # vars from blob0
    VAR3: potato1  # vars from blob1

- k1: v1
  vars: &splat1
    VAR1: spud    # vars from blob1

- k: v
  extra:          # merged vars from blob0 + blob1
    <<: *splat0
    <<: *splat1
```

output:
```
[{'k0': 'v0', 'vars': {'VAR0': 'potato', 'VAR3': 'potato1'}}, {'k1': 'v1', 'vars': {'VAR1': 'spud'}}, {'k': 'v', 'extra': {'VAR0': 'potato', 'VAR3': 'potato1', 'VAR1': 'spud'}}]
```

invalid input:
```
- k0: v0
  vars: &splat0
    VAR0: potato  # vars from blob0
    VAR1: potato1  # vars from blob1

- k1: v1
  vars: &splat1
    VAR1: spud    # vars from blob1

- k: v
  extra:          # merged vars from blob0 + blob1
    <<: *splat0
    <<: *splat1
```

output:
```
Duplicate key detected: Duplicate key 'VAR1' found
Each key must be unique in YAML mapping
  in "demo.yaml", line 8, column 5
```
